src/Cli/dotnet/commands/dotnet-run/VirtualProjectBuildingCommand.cs

```diff
@@ -139,86 +178,443 @@ private ProjectInstance CreateProjectInstance(
+    {
+        WriteProjectFile(writer, directives, virtualProjectFile: false, targetFilePath: null);
+    }
 
-    public static string GetNonVirtualProjectFileText()
+    private static void WriteProjectFile(TextWriter writer, ImmutableArray<CSharpDirective> directives, bool virtualProjectFile, string? targetFilePath)
     {
-        return $"""
-            <Project Sdk="Microsoft.NET.Sdk">
+        int processedDirectives = 0;
 
+        var sdkDirectives = directives.OfType<CSharpDirective.Sdk>();
+        var propertyDirectives = directives.OfType<CSharpDirective.Property>();
+        var packageDirectives = directives.OfType<CSharpDirective.Package>();
+
+        string sdkValue = "Microsoft.NET.Sdk";
+
+        if (sdkDirectives.FirstOrDefault() is { } firstSdk)
+        {
+            sdkValue = firstSdk.ToSlashDelimitedString();
+            processedDirectives++;
+        }
+
+        if (virtualProjectFile)
+        {
+            writer.WriteLine($"""
+                <Project>
+
+                  <!-- We need to explicitly import Sdk props/targets so we can override the targets below. -->
+                  <Import Project="Sdk.props" Sdk="{EscapeValue(sdkValue)}" />
+                """);
+        }
+        else
+        {
+            writer.WriteLine($"""
+                <Project Sdk="{EscapeValue(sdkValue)}">
+
+                """);
+        }
+
+        foreach (var sdk in sdkDirectives.Skip(1))
+        {
+            if (virtualProjectFile)
+            {
+                writer.WriteLine($"""
+                      <Import Project="Sdk.props" Sdk="{EscapeValue(sdk.ToSlashDelimitedString())}" />
+                    """);
+            }
+            else if (sdk.Version is null)
+            {
+                writer.WriteLine($"""
+                      <Sdk Name="{EscapeValue(sdk.Name)}" />
+                    """);
+            }
+            else
+            {
+                writer.WriteLine($"""
+                      <Sdk Name="{EscapeValue(sdk.Name)}" Version="{EscapeValue(sdk.Version)}" />
+                    """);
+            }
+
+            processedDirectives++;
+        }
+
+        if (processedDirectives > 1)
+        {
+            writer.WriteLine();
+        }
+
+        // Kept in sync with the default `dotnet new console` project file (enforced by `DotnetProjectAddTests.SameAsTemplate`).
+        writer.WriteLine($"""
               <PropertyGroup>
-            {CommonProjectProperties}
+                <OutputType>Exe</OutputType>
+                <TargetFramework>net10.0</TargetFramework>
+                <ImplicitUsings>enable</ImplicitUsings>
+                <Nullable>enable</Nullable>
               </PropertyGroup>
+            """);
+
+        if (virtualProjectFile)
+        {
+            writer.WriteLine("""
+
+                  <PropertyGroup>
+                    <EnableDefaultItems>false</EnableDefaultItems>
```

Comment:
Can we pre-loop the property directives to see if this is being explicitly set in the C# file and omit this in that case? Or we  happy just relying on evaluation order given that we're writing the property directives after this?

Reply:
Evaluation order should work. Do you have any concerns about it?

Reply:
I'm happy relying on evaluation order for now. We could be smarter for the other properties as well. For example, if user has `#:property Nullable=disable`, we could replace the default `<Nullable>enable</Nullable>`. But that's tricky, because if user did `#:property Nullable=$(Nullable);something` for some reason, we probably shouldn't replace the default...